### PR TITLE
Allow `__getitem__` indices that have a `to_dask_array` method

### LINKED ISSUE #677: Allow `__getitem__` indices that have a `to_dask_array` method
It would be useful to allow `cf.Data` and other objects (such as `cf.DimensionCoordinate`) to be used as indices within cf-python.

E.g. so we could do a northern hemisphere subspace like this:

```python
>>> g = f[f.dimension_coordinate('Y') > 0, ...]
```

Note: you can already do this with ` g = f.subspace(Y=cf.gt(0))`, but there's no reason to preculde the `[]` approach, which currently does not work (but did pre 3.14.0).

PR to follow.

Thanks to @sws06mas for bring this to my attention.

### PR DESCRIPTION
**Pull request by davidhassell**
Fixes #677 

Plus a bit of linting on some previous, unconnected work.

### CODE REVIEW

**Review by sadielbartholomew — COMMENTED** [review]
Works to re-implement the feature at hand, but ideally we can add to this PR to:
* add a little testing of the indices on the Field side in `test_Field__getitem__` as well; and
* also (in testing) cover cases of objects with masking as indices since masked arrays can be pesky;
* add documentation to highlight the possibility of using cf objects as indices, e.g. adding to the ['Assignment by index' section](https://ncas-cms.github.io/cf-python/tutorial.html#assignment-by-index) of the tutorial.

I've raised one minor comment and also have one question:
* should cf objects with a `to_dask_array` method now be valid inputs to the `indices` methods of `Field` and `Domain` (etc.) because they don't seem to work with those at the moment (but maybe they shouldn't and I am missing something!).

File: Changelog.rst [hunk 1]
```diff
@@ -7,6 +7,8 @@ version ?.?.?
   (https://github.com/NCAS-CMS/cf-python/issues/661)
 * New keyword parameter to `cf.aggregate`: ``cells``
   (https://github.com/NCAS-CMS/cf-python/issues/452)
+* Allow `__getitem__` indices that have a `to_dask_array` method
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply Changelog.rst]: https://github.com/NCAS-CMS/cf-python/pull/678/commits/2aa45ecacaac67f6a98efaf01711d0389e6613b6

As a firm adherent to the many-universes hypothesis, I can rest happy knowing that the obfuscated developers version lives on somewhere :)

[the remark this answers is not in the document]
Comment (davidhassell) [reply Changelog.rst]: I'm looking at the improving the tests and docs as you have suggested ...

**Comment by davidhassell** [comment]
Hi @sadielbartholomew,

> add a little testing of the indices on the Field side in test_Field__getitem__ as well; and also (in testing) cover cases of objects with masking as indices since masked arrays can be pesky;

https://github.com/NCAS-CMS/cf-python/pull/678/commits/5a28c946694a673fbd29b692947d64fad6564440

>    add documentation to highlight the possibility of using cf objects as indices, e.g. adding to the ['Assignment by index' section](https://ncas-cms.github.io/cf-python/tutorial.html#assignment-by-index) of the tutorial.

https://github.com/NCAS-CMS/cf-python/pull/678/commits/6fb4e472250ec06d8865343038f235d56097c61f

File: docs/source/tutorial.rst [hunk 2]
```diff
@@ -1427,8 +1427,25 @@ the only differences being:
                    : longitude(5) = [157.5, ..., -22.5] degrees_east
                    : time(1) = [2019-01-01 00:00:00]
 
+An index that is a list of integers may also be provided as a `numpy`
+array, a `dask` array, a `cf.Data` instance, or a metadata construct
+that has data:
+
+.. code-block:: python
+   :caption: *Create a new field construct for longitudes greater than
+             180.*
```
Comment (sadielbartholomew) [inline docs/source/tutorial.rst:1436]:
```suggestion
             180 degrees east.*
```

Simply for good practice with always including units, even if it is obvious from context :slightly_smiling_face:

**Review by sadielbartholomew — APPROVED** [review]
Previous feedback addressed well. Thanks David. I've done a final sanity check and all tests pass locally and everything's good. Please merge!

**Pull request merged**